libraries/chain/controller.cpp

```diff
@@ -3999,15 +3986,58 @@ struct controller_impl {
       bsp->verify_qc(qc_proof);
    }
 
+   // Verify a Legacy block does not have a QC block extension.
+   // (It can have a finality block header extension during transition to
+   // Savanna)
+   void verify_legacy_block_exts( const signed_block_ptr& b, const block_header_state_legacy& prev ) {
+      uint32_t block_num = b->block_num();
+
+      auto block_exts = b->validate_and_extract_extensions();
+      auto qc_ext_id = quorum_certificate_extension::extension_id();
+      bool qc_extension_present = block_exts.count(qc_ext_id) != 0;
+      EOS_ASSERT( !qc_extension_present,
+                  block_validate_exception,
+                  "Legacy block #${b} includes a QC block extension",
+                  ("b", block_num) );
+
```

Comment:
We should also check that the previous block's header satisfies `is_proper_svnn_block() == false`.

Reply:
And for that matter, we could also check that the current block satisfied `is_proper_svnn_block() == false`.